File: mobile2/lib/services/notification_service.dart
import 'dart:async';
import 'dart:convert';
import 'package:flutter/foundation.dart';
import 'package:web_socket_channel/web_socket_channel.dart';
import 'package:web_socket_channel/status.dart' as status;

/// Event types for real-time notifications
enum NotificationEventType {
  meetingStatusChanged('meeting.status_changed'),
  meetingUpdated('meeting.updated'),
  recordingStarted('recording.started'),
  recordingCompleted('recording.completed'),
  recordingFailed('recording.failed'),
  transcriptionStarted('transcription.started'),
  transcriptionCompleted('transcription.completed'),
  transcriptionFailed('transcription.failed'),
  summaryStarted('summary.started'),
  summaryCompleted('summary.completed'),
  summaryFailed('summary.failed'),
  compositeVideoStarted('composite_video.started'),
  compositeVideoCompleted('composite_video.completed'),
  compositeVideoFailed('composite_video.failed'),
  systemConnected('system.connected'),
  systemPong('system.pong');

  const NotificationEventType(this.value);
  final String value;

  static NotificationEventType? fromString(String value) {
    for (var type in NotificationEventType.values) {
      if (type.value == value) return type;
    }
    return null;
  }
}

/// Real-time notification model
class Notification {
  final String id;
  final NotificationEventType type;
  final String entityType;
  final String entityId;
  final String? meetingId;
  final String? userId;
  final Map<String, dynamic>? changedFields;
  final DateTime timestamp;
  final String? message;

  Notification({
    required this.id,
    required this.type,
    required this.entityType,
    required this.entityId,
    this.meetingId,
    this.userId,
    this.changedFields,
    required this.timestamp,
    this.message,
  });

  factory Notification.fromJson(Map<String, dynamic> json) {
    return Notification(
      id: json['id'] as String,
      type: NotificationEventType.fromString(json['type'] as String) ??
          NotificationEventType.systemConnected,
      entityType: json['entity_type'] as String? ?? '',
      entityId: json['entity_id'] as String? ?? '',
      meetingId: json['meeting_id'] as String?,
      userId: json['user_id'] as String?,
      changedFields: json['changed_fields'] as Map<String, dynamic>?,
      timestamp: json['timestamp'] != null
          ? DateTime.parse(json['timestamp'] as String)
          : DateTime.now(),
      message: json['message'] as String?,
    );
  }
}

/// Callback for notification events
typedef NotificationHandler = void Function(Notification notification);

/// WebSocket-based real-time notification service
class NotificationService {
  final String baseUrl;
  final String Function() getToken;

  WebSocketChannel? _channel;
  StreamSubscription? _subscription;
  Timer? _reconnectTimer;
  Timer? _pingTimer;
  int _reconnectAttempts = 0;
  static const int _maxReconnectAttempts = 10;
  static const int _baseReconnectDelay = 1000; // milliseconds
  bool _isIntentionallyClosed = false;

  final List<NotificationHandler> _handlers = [];

  NotificationService({
    required this.baseUrl,
    required this.getToken,
  });

  /// Connect to the WebSocket endpoint
  Future<void> connect() async {
    if (_channel != null) {
      debugPrint('[NotificationService] Already connected');
      return;
    }

    final token = getToken();
    if (token.isEmpty) {
      debugPrint('[NotificationService] ❌ No auth token available');
      return;
    }

    _isIntentionallyClosed = false;

    // Build WebSocket URL
    final protocol = baseUrl.startsWith('https') ? 'wss' : 'ws';
    final wsUrl = '$protocol://${baseUrl.replaceAll(RegExp(r'^https?://'), '')}/api/v1/notifications/ws';

    debugPrint('[NotificationService] Connecting to: $wsUrl');

    try {
      // Note: Authorization header doesn't work with WebSocket in browsers
      // The server should handle auth through a separate mechanism or query params
      _channel = WebSocketChannel.connect(
        Uri.parse(wsUrl),
      );

      debugPrint('[NotificationService] ✅ Connected to notification service');
      _reconnectAttempts = 0;

      // Send authentication message
      _sendMessage({'type': 'auth', 'token': token});

      // Listen to messages
      _subscription = _channel!.stream.listen(
        _handleMessage,
        onError: (error) {
          debugPrint('[NotificationService] ❌ Error: $error');
          _handleDisconnect();
        },
        onDone: () {
          debugPrint('[NotificationService] 🔌 Disconnected');
          _handleDisconnect();
        },
      );

      // Start ping timer to keep connection alive
      _startPingTimer();
    } catch (error) {
      debugPrint('[NotificationService] ❌ Failed to connect: $error');
      _scheduleReconnect();
    }
  }

  /// Disconnect from the WebSocket
  void disconnect() {
    debugPrint('[NotificationService] Disconnecting...');
    _isIntentionallyClosed = true;

    _stopPingTimer();
    _stopReconnectTimer();

    _subscription?.cancel();
    _subscription = null;

    _channel?.sink.close(status.goingAway);
    _channel = null;

    _reconnectAttempts = 0;
  }

  /// Subscribe to notifications
  void Function() subscribe(NotificationHandler handler) {
    _handlers.add(handler);

    // Return unsubscribe function
    return () {
      _handlers.remove(handler);
    };
  }

  /// Check if connected
  bool get isConnected => _channel != null;

  /// Send a ping message
  void ping() {
    _sendMessage({'type': 'ping'});
  }

  /// Handle incoming messages
  void _handleMessage(dynamic data) {
    try {
      final json = jsonDecode(data as String) as Map<String, dynamic>;
      final notification = Notification.fromJson(json);

      debugPrint('[NotificationService] 📨 Received: ${notification.type.value}');

      // Notify all handlers
      for (final handler in _handlers) {
        try {
          handler(notification);
        } catch (error) {
          debugPrint('[NotificationService] Error in handler: $error');
        }
      }
    } catch (error) {
      debugPrint('[NotificationService] Failed to parse notification: $error');
    }
  }

  /// Handle disconnection
  void _handleDisconnect() {
    _subscription?.cancel();
    _subscription = null;
    _channel = null;
    _stopPingTimer();

    if (!_isIntentionallyClosed) {
      _scheduleReconnect();
    }
  }

  /// Schedule reconnection attempt
  void _scheduleReconnect() {
    if (_reconnectAttempts >= _maxReconnectAttempts) {
      debugPrint('[NotificationService] ❌ Max reconnect attempts reached');
      return;
    }

    if (_reconnectTimer != null) {
      return; // Already scheduled
    }

    final delay = Duration(
      milliseconds: (_baseReconnectDelay *
              (1 << _reconnectAttempts).clamp(1, 30))
          .toInt(),
    );

    debugPrint(
        '[NotificationService] Reconnecting in ${delay.inSeconds}s (attempt ${_reconnectAttempts + 1}/$_maxReconnectAttempts)');

    _reconnectTimer = Timer(delay, () {
      _reconnectTimer = null;
      _reconnectAttempts++;
      connect();
    });
  }

  /// Stop reconnect timer
  void _stopReconnectTimer() {
    _reconnectTimer?.cancel();
    _reconnectTimer = null;
  }

  /// Start ping timer
  void _startPingTimer() {
    _pingTimer = Timer.periodic(const Duration(seconds: 30), (timer) {
      if (_channel != null) {
        ping();
      }
    });
  }

  /// Stop ping timer
  void _stopPingTimer() {
    _pingTimer?.cancel();
    _pingTimer = null;
  }

  /// Send a message to the server
  void _sendMessage(Map<String, dynamic> data) {
    if (_channel != null) {
      _channel!.sink.add(jsonEncode(data));
    }
  }

  /// Dispose the service
  void dispose() {
    disconnect();
  }
}

File: front/user-portal/src/services/notificationService.ts
/**
 * Real-time Notification Service using WebSocket
 * Connects to /api/v1/notifications/ws and receives real-time updates
 */

export type EventType =
  | 'meeting.status_changed'
  | 'meeting.updated'
  | 'recording.started'
  | 'recording.completed'
  | 'recording.failed'
  | 'transcription.started'
  | 'transcription.completed'
  | 'transcription.failed'
  | 'summary.started'
  | 'summary.completed'
  | 'summary.failed'
  | 'composite_video.started'
  | 'composite_video.completed'
  | 'composite_video.failed'
  | 'system.connected'
  | 'system.pong';

export interface Notification {
  id: string;
  type: EventType;
  entity_type: string; // meeting, recording, transcription, summary
  entity_id: string;
  meeting_id?: string;
  user_id?: string;
  changed_fields?: Record<string, any>;
  timestamp: string;
  message?: string;
}

export type NotificationHandler = (notification: Notification) => void;

export class NotificationService {
  private ws: WebSocket | null = null;
  private handlers: Set<NotificationHandler> = new Set();
  private reconnectTimeout: NodeJS.Timeout | null = null;
  private reconnectAttempts = 0;
  private maxReconnectAttempts = 10;
  private baseReconnectDelay = 1000; // 1 second
  private isIntentionallyClosed = false;

  constructor(private baseUrl: string, private getToken: () => string | null) {}

  /**
   * Connect to the WebSocket endpoint
   */
  connect(): void {
    if (this.ws?.readyState === WebSocket.OPEN) {
      console.log('[NotificationService] Already connected');
      return;
    }

    const token = this.getToken();
    if (!token) {
      console.error('[NotificationService] No auth token available');
      return;
    }

    this.isIntentionallyClosed = false;

    // Build WebSocket URL
    const protocol = this.baseUrl.startsWith('https') ? 'wss' : 'ws';
    const wsUrl = `${protocol}://${this.baseUrl.replace(/^https?:\/\//, '')}/api/v1/notifications/ws`;

    console.log('[NotificationService] Connecting to:', wsUrl);

    try {
      this.ws = new WebSocket(wsUrl);

      // Send auth token after connection
      this.ws.onopen = () => {
        console.log('[NotificationService] ✅ Connected to notification service');
        this.reconnectAttempts = 0;

        // Send authentication in first message
        this.send({ type: 'auth', token });
      };

      this.ws.onmessage = (event) => {
        try {
          const notification: Notification = JSON.parse(event.data);
          console.log('[NotificationService] 📨 Received notification:', notification);

          // Notify all registered handlers
          this.handlers.forEach((handler) => {
            try {
              handler(notification);
            } catch (error) {
              console.error('[NotificationService] Error in notification handler:', error);
            }
          });
        } catch (error) {
          console.error('[NotificationService] Failed to parse notification:', error);
        }
      };

      this.ws.onerror = (error) => {
        console.error('[NotificationService] ❌ WebSocket error:', error);
      };

      this.ws.onclose = (event) => {
        console.log('[NotificationService] 🔌 Disconnected:', event.code, event.reason);
        this.ws = null;

        // Attempt to reconnect if not intentionally closed
        if (!this.isIntentionallyClosed) {
          this.scheduleReconnect();
        }
      };
    } catch (error) {
      console.error('[NotificationService] Failed to create WebSocket:', error);
      this.scheduleReconnect();
    }
  }

  /**
   * Disconnect from the WebSocket
   */
  disconnect(): void {
    console.log('[NotificationService] Disconnecting...');
    this.isIntentionallyClosed = true;

    if (this.reconnectTimeout) {
      clearTimeout(this.reconnectTimeout);
      this.reconnectTimeout = null;
    }

    if (this.ws) {
      this.ws.close();
      this.ws = null;
    }

    this.reconnectAttempts = 0;
  }

  /**
   * Schedule reconnection attempt with exponential backoff
   */
  private scheduleReconnect(): void {
    if (this.reconnectAttempts >= this.maxReconnectAttempts) {
      console.error('[NotificationService] Max reconnect attempts reached');
      return;
    }

    if (this.reconnectTimeout) {
      return; // Already scheduled
    }

    const delay = Math.min(
      this.baseReconnectDelay * Math.pow(2, this.reconnectAttempts),
      30000 // Max 30 seconds
    );

    console.log(
      `[NotificationService] Reconnecting in ${delay}ms (attempt ${this.reconnectAttempts + 1}/${this.maxReconnectAttempts})`
    );

    this.reconnectTimeout = setTimeout(() => {
      this.reconnectTimeout = null;
      this.reconnectAttempts++;
      this.connect();
    }, delay);
  }

  /**
   * Subscribe to notifications
   */
  subscribe(handler: NotificationHandler): () => void {
    this.handlers.add(handler);

    // Return unsubscribe function
    return () => {
      this.handlers.delete(handler);
    };
  }

  /**
   * Send a message to the server
   */
  private send(data: any): void {
    if (this.ws?.readyState === WebSocket.OPEN) {
      this.ws.send(JSON.stringify(data));
    }
  }

  /**
   * Check if connected
   */
  isConnected(): boolean {
    return this.ws?.readyState === WebSocket.OPEN;
  }

  /**
   * Send ping to keep connection alive
   */
  ping(): void {
    this.send({ type: 'ping' });
  }
}

// Singleton instance
let notificationService: NotificationService | null = null;

/**
 * Get or create the notification service instance
 */
export function getNotificationService(
  baseUrl: string,
  getToken: () => string | null
): NotificationService {
  if (!notificationService) {
    notificationService = new NotificationService(baseUrl, getToken);
  }
  return notificationService;
}

/**
 * React hook for using notifications
 */
export function useNotificationService(
  baseUrl: string,
  getToken: () => string | null,
  handler?: NotificationHandler
): NotificationService {
  const service = getNotificationService(baseUrl, getToken);

  // Subscribe to notifications if handler provided
  if (handler) {
    const unsubscribe = service.subscribe(handler);
    // Note: In a real React hook, this should be wrapped in useEffect with cleanup
    return service;
  }

  return service;
}
